src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -405,9 +370,7 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
   if (result != nullptr) {
     // Allocation successful, bump stats:
     if (req.is_mutator_alloc()) {
-      assert(req.is_young(), "Mutator allocations always come from young generation.");
-      generation->increase_used(size * HeapWordSize);
-      increase_used(size * HeapWordSize);
+      increase_used(req.actual_size() * HeapWordSize);
```

Comment:
There is leakage of abstraction happening here. I'd prefer if the adjustments of heap use occurred when the space was allocated by the region. That avoids accounting errors because of piecemeal adjustments scattered along the allocation paths here.

Reply:
Right - your withdrawn comment is just what I was going for with this change. `ShenandoahFreeSet` maintains its own notion of `used`, which has similar but subtly different rules for the accounting that happens for generations and the heap. Perhaps they could be merged at some point.